x/distribution/keeper/calculation.go

```diff
@@ -0,0 +1,159 @@
+
+	"github.com/cosmos/cosmos-sdk/x/distribution/types"
+)
+
+// initialize rewards for a new validator
+func (k Keeper) initializeValidator(ctx sdk.Context, val sdk.Validator) {
+	// set initial historical rewards (period 0)
+	k.SetValidatorHistoricalRewards(ctx, val.GetOperator(), 0, types.ValidatorHistoricalRewards{})
+
+	// set current rewards (starting at period 1)
+	k.SetValidatorCurrentRewards(ctx, val.GetOperator(), types.ValidatorCurrentRewards{
+		Rewards: sdk.DecCoins{},
+		Period:  1,
+	})
+
+	// set accumulated commission
+	k.SetValidatorAccumulatedCommission(ctx, val.GetOperator(), types.ValidatorAccumulatedCommission{})
+}
+
+// increment validator period, returning the period just ended
+func (k Keeper) incrementValidatorPeriod(ctx sdk.Context, val sdk.Validator) uint64 {
+	// Fetch current rewards
+	rewards := k.GetValidatorCurrentRewards(ctx, val.GetOperator())
+
+	// Calculate current ratio
+	var current sdk.DecCoins
+	if val.GetTokens().IsZero() {
+
+		// can't calculate ratio for zero-token validators
+		// ergo we instead add to the community pool
+		feePool := k.GetFeePool(ctx)
+		outstanding := k.GetOutstandingRewards(ctx)
+		feePool.CommunityPool = feePool.CommunityPool.Plus(rewards.Rewards)
+		outstanding = outstanding.Minus(rewards.Rewards)
+		k.SetFeePool(ctx, feePool)
+		k.SetOutstandingRewards(ctx, outstanding)
+
+		current = sdk.DecCoins{}
+	} else {
+		current = rewards.Rewards.QuoDec(sdk.NewDecFromInt(val.GetTokens()))
+	}
+
+	// Fetch historical rewards for last period
+	historical := k.GetValidatorHistoricalRewards(ctx, val.GetOperator(), rewards.Period-1)
+
+	// Set new historical rewards
+	k.SetValidatorHistoricalRewards(ctx, val.GetOperator(), rewards.Period, historical.Plus(current))
+
+	// Set current rewards, incrementing period by 1
+	newPeriod := rewards.Period + 1
+	k.SetValidatorCurrentRewards(ctx, val.GetOperator(), types.ValidatorCurrentRewards{
+		Rewards: sdk.DecCoins{},
+		Period:  newPeriod,
+	})
+
+	return rewards.Period
+}
+
+// initialize starting info for a new delegation
+func (k Keeper) initializeDelegation(ctx sdk.Context, val sdk.ValAddress, del sdk.AccAddress) {
+	// period has already been incremented
+	period := k.GetValidatorCurrentRewards(ctx, val).Period
+	validator := k.stakingKeeper.Validator(ctx, val)
+	delegation := k.stakingKeeper.Delegation(ctx, del, val)
+	// calculate delegation staking in tokens
+	// TODO need to make sure this truncates instead of rounding
+	stake := delegation.GetShares().Mul(validator.GetDelegatorShareExRate())
+	k.SetDelegatorStartingInfo(ctx, val, del, types.DelegatorStartingInfo{
+		PreviousPeriod: period - 1,
+		Stake:          stake,
+		Height:         uint64(ctx.BlockHeight()),
+	})
+}
+
+// calculate the rewards accrued by a delegation between two periods
+func (k Keeper) calculateDelegationRewardsBetween(ctx sdk.Context, val sdk.Validator, startingPeriod uint64, endingPeriod uint64, staking sdk.Dec) (rewards sdk.DecCoins) {
```

Comment:
```suggestion
func (k Keeper) calculateDelegationRewardsBetween(ctx sdk.Context, val sdk.Validator, startingPeriod, endingPeriod uint64, staking sdk.Dec) (rewards sdk.DecCoins) {
```

Can we also make this a multi-line signature? It's pretty long as-is.